src/loam/allocation.rs

```diff
@@ -346,174 +516,164 @@ ascent! {
-            //assert!(prog.cons_mem_is_contiguous());
-
-            prog
+    fn wide_ptr(tag: LE, digest: [LE; 8]) -> WidePtr {
+        WidePtr(Wide::widen(tag), Wide(digest))
+    }
+
+    fn read_wideptr(zstore: &mut ZStore<BabyBear, LurkChip>, src: &str) -> WidePtr {
+        let ZPtr { tag, digest } = zstore.read(src).unwrap();
+        wide_ptr(tag.elt(), digest)
+    }
+
+    fn test_aux0(
+        zstore: &ZStore<BabyBear, LurkChip>,
+        input: WidePtr,
+        expected_output: WidePtr,
+    ) -> AllocationProgram {
+        let mut prog = AllocationProgram::default();
+
+        prog.import_zstore(zstore);
+        prog.input_expr = vec![(input,)];
+        prog.run();
+
+        println!("{}", prog.relation_sizes_summary());
+
+        assert_eq!(expected_output, prog.output_expr[0].0);
+        prog
+    }
+
+    fn test_aux(input: &str, expected_output: &str) -> AllocationProgram {
+        let mut zstore = lurk_zstore();
+        let input = read_wideptr(&mut zstore, input);
+        let expected_output = read_wideptr(&mut zstore, expected_output);
+        test_aux0(&zstore, input, expected_output)
+    }
+
+    fn test_aux1(input: &str, expected_output: WidePtr) -> AllocationProgram {
+        let mut zstore = lurk_zstore();
+        let input = read_wideptr(&mut zstore, input);
+        test_aux0(&zstore, input, expected_output)
+    }
+
+    fn test_distilled(input: &str, expected_output: &str, bad_input_output: Option<(&str, &str)>) {
+        // Run the first phase
+        let mut zstore = lurk_zstore();
+        let input = read_wideptr(&mut zstore, input);
+        let expected_output = read_wideptr(&mut zstore, expected_output);
+        let original_program = test_aux0(&zstore, input, expected_output);
+
+        // Start the second phase
+        let mut prog = DistilledAllocationProgram::default();
+
+        // Move the hashes (stored in the allocator) and input over
+        prog.allocator = original_program.allocator.clone();
+        prog.input_expr = original_program.input_expr.clone();
+
+        // transfer over the memory (assume it's been distilled)
+        let raw_memory = original_program.export_memory();
+        let mut store = Store::default();
+        let options = DistillationOptions::new().with_summary(0.9);
+        let memory = raw_memory.distill_with_store(&mut store, &options);
+
+        // Determine whether we want to use the intended in/out, or the bad in/out
+        let (input_ptr, output_expr) = if let Some((bad_input, bad_output)) = bad_input_output {
+            let bad_zptr = zstore.read(bad_input).expect("failed to read");
+            (
+                store.zptr_ptr(&bad_zptr).unwrap(),
+                vec![(read_wideptr(&mut zstore, bad_output),)],
+            )
+        } else {
+            let input_zptr = prog.input_expr[0].0.to_zptr();
+            (
+                store.zptr_ptr(&input_zptr).unwrap(),
+                original_program.output_expr,
```

Comment:
What is actually being tested here? Is it the possibility that the `unwrap`s might fail? I would expect to see some assertion performed using these results.

Reply:
I understand what's going on now, but it's hard to follow. I think a little documentation of how this is meant to be used would help.